Direct byte code generation

pontusmelke opened the pull request:
Instead of relying on `javax.tools` and `javac` to generate the byte code from java source, we instead produce byte code directly. The main advantages are:
- `tools.jar` doesn't need to be in the class path, meaning that it works on all `jres` and `jdks`
- Faster, 7-8 times faster code generation

The commit history is intentionally pretty long since it is a rather big PR but there is some back-and-forth changes during the development.
Changes include:
- Not longer relying on `try-with-resources` in the generated code
- Clean up in the SPIs for code generation
- Not relying on automatic boxing and unboxing
- Fixed a lot of small bugs and inefficiencies

pontusmelke committed:
ByteCode support for Classes and Methods

pontusmelke committed:
Bump asm version

pontusmelke committed:
Made it configurable to use Byte Code generation

pontusmelke committed:
Support for class fields

pontusmelke committed:
Support for calling super, put, load

- always generate a default constructor when there are none
- call super constructor
- support for put field
- support for load

pontusmelke committed:
Call default super constructor when necessary

If not explicit call to a super constructor is made as the top of the
constructor, automatically generate byte code for calling the default
constructor of the super class.

pontusmelke committed:
Support for get field and return expression

pontusmelke committed:
Support for static get and put

In implementing this also fixed

- add `newArray` so vararg method can be used
 - fixed bug in handling of type names of reference array types

pontusmelke committed:
Support for inner classes

pontusmelke committed:
Support throwing parameterized checked exception

pontusmelke committed:
Support local variable assignment

pontusmelke committed:
Support for while loops

- Broke out classes to files
- Introduced blocks to handle different endBlock instructions

pontusmelke committed:
Support for if

pontusmelke committed:
Support for or

pontusmelke committed:
Support try-catch-finally

This more or less rewrites how we do try-catch-finally. From an API point-of-view it looks
the same but behind the scenes we actually create the entire try-catch-finally in one go.

pontusmelke committed:
Support for throwing exceptions

pontusmelke committed:
Support for declare and assign

pontusmelke committed:
Support for foreach

Rewrite the foreach to just be syntactic sugar for a while loop over the corresponding
iterator since that is the way the byte code will look anyhow.

pontusmelke committed:
Support for not

pontusmelke committed:
Support for ternary operator

pontusmelke committed:
Support for equality and greater than comparison

pontusmelke committed:
Support for addition and subtraction

pontusmelke committed:
Enable byte code generation in birk

pontusmelke committed:
Fixed proper boxing and addition

pontusmelke committed:
Fixed subtraction

pontusmelke committed:
Fixed multiplication

pontusmelke committed:
Fixed division

pontusmelke committed:
Fixed the modulos operator

pontusmelke committed:
Cleaned up arithmetic operations

pontusmelke committed:
Changed equality to consider type

pontusmelke committed:
Changed greater than to consider type

pontusmelke committed:
Clean up around eq

pontusmelke committed:
Refactor and more tests for CodeGeneratedQueryStructure

pontusmelke committed:
Cleaned up tryCatch code generation

pontusmelke committed:
Simplified configuration of CodeGenerator

pontusmelke committed:
Fixed type errors in hash joins

pontusmelke committed:
Added ifNot as a more efficient if(not(test))

pontusmelke committed:
Added more efficient null checking primitives

pontusmelke committed:
Fixed bug in predicate handlings

pontusmelke committed:
Fixed bug when materializing relationships

pontusmelke committed:
Properly close when exceptions are thrown

pontusmelke committed:
Enabled tracing again

pontusmelke committed:
Capture both CypherType and RepresentationType

Refactor: introduce representation type

- Capture both CypherType and RepresentationType in CodeGenExpression
- Fixed failing tests
- Some clean up

pontusmelke committed:
Clean up and comments

systay commented:
Wow... This is really great stuff, @pontusmelke!

systay commented on community/codegen/src/main/java/org/neo4j/codegen/BaseExpressionVisitor.java:
```diff
@@ -0,0 +1,188 @@
+/**
+ * Default implementation of {@link ExpressionVisitor}
+ */
+public abstract class BaseExpressionVisitor implements ExpressionVisitor
```
Maybe add a comment for why we need an abstract class that implements all interface methods with empty implementations?

systay commented on community/codegen/src/main/java/org/neo4j/codegen/bytecode/ByteCodeExpressionVisitor.java:
```diff
@@ -0,0 +1,542 @@
+import org.neo4j.codegen.MethodReference;
+import org.neo4j.codegen.TypeReference;
+
+import static org.neo4j.codegen.ByteCodeUtils.byteCodeName;
```
Why a static dependency instead of DI through the constructor?

pontusmelke replied:
ByteCodeUtils just contains static helper methods, it is not an instantiable class.

systay replied:
I know. That is kind of my question - why is it a static class instead of a normal class that we can mock out or change?

pontusmelke replied:
Sure we could do that. I am not sure I see the point though, since input is typically a `TypeReference` say a `String`-reference and output is the "bytecode type", for example  `Ljava/lang/String`. To me mocking this would to me be a little like mocking `java.lang.Math.abs` - you could do it but what is the point, when do you want to mock a function to return the incorrect output given an input?

thobe commented:
I'll have a look at this next week.

pontusmelke committed:
Remove automatic call to default super constructor

Though convenient  to not have to call `super()` the actual benefit was not
worth it in terms of the extra complexity it added. The actual implementation
though working currently was probably incorrect and could lead to errors if
more operations added to the code generation framework.

henriknyman commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/codegen/LogicalPlanConverter.scala:
```diff
@@ -74,7 +74,7 @@ object LogicalPlanConverter {
       val vars = columns.map {
         case (name, expr) =>
-          val variable = Variable(context.namer.newVarName(), expr.cypherType(context), expr.nullable(context))
+          val variable = Variable(context.namer.newVarName(), symbols.CTAny, expr.nullable(context))
```
Why do we not use the expression type anymore?

pontusmelke replied:
When we are projecting we are basically erasing the type information, since we are doing

```
Object res = ....
row.set("foo", res);
```

I think we should improve this and keep the type information, but maybe we can do this as a separate piece of work?

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/ByteCodeUtils.java:
```diff
@@ -0,0 +1,233 @@
+
+import java.util.List;
+
+public final class ByteCodeUtils
```
A lot of the functionality in this class already exists in `org.objectweb.asm.Type`.
It might still make sense to have this class, in order to own our interfaces, but the implementation could perhaps. be made by delegating to `org.objectweb.asm.Type`.

pontusmelke replied:
That sounds like a great idea. I just need to take time to figure out what all the methods in `org.objectweb.asm.Type` and which of them to use.

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/MethodDeclaration.java:
```diff
@@ -172,10 +211,63 @@ public TypeReference declaringClass()
+        String newName = name();
+        boolean newIsConstrucor = isConstructor();
+
+        return new MethodDeclaration( owner, newParameters, newExceptions, modifiers, typeParameters )
```
Since this is declared in a method, it has a reference to the object that it was instantiated from (whether it uses it or not). I don't know if it is harmful (haven't done the full analysis yet), but it might keep that object alive for longer than needed.

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/ExpressionTemplate.java:
```diff
@@ -219,16 +222,17 @@ protected void templateAccept( CodeBlock method, ExpressionVisitor visitor )
     }
 
-    static ExpressionTemplate invokeSuperConstructor( final ExpressionTemplate[] parameters )
+    //TODO I am not crazy about the way type parameters are sent here
```
Would you care to elaborate?
1. What is it you don't like about this way?
2. Do you have any idea on how you would like it to be done?

Could it possibly be done using only a single `parameters` array, if the `ExpressionTemplate` has a type? (if a different type than the actual evaluated type of the expression is used in the signature of the constructor, we could have some sort of `expr.asType( typeRef )` method)

pontusmelke replied:
Having two arrays that are required to have the same size is kind of smelly and clunky to use. I think we can do something like you propose but it is kind of a big change I think. Furthermore this method is currently only used from one test and not used at all in the production code at all so it is probably not the highest priority to clean up.

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/bytecode/Block.java:
```diff
@@ -17,22 +17,11 @@
-    {
-        return emitFinally();
-    }
+interface Block extends Opcodes
```
I'm not a fan of bringing in constants by implementing an interface. `import static  org.objectweb.asm.Opcodes.*` would be preferable IMO.

pontusmelke committed:
Fixes from code review

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/bytecode/ClassByteCodeWriter.java:
```diff
@@ -0,0 +1,136 @@
+
+    ClassByteCodeWriter( TypeReference type, TypeReference base, TypeReference[] interfaces )
+    {
+        this.classWriter = new ClassWriter( ClassWriter.COMPUTE_FRAMES );
```
Does `COMPUTE_FRAMES` imply `COMPUTE_MAXS`, or should we maybe pass `COMPUTE_FRAMES | COMPUTE_MAXS` in here?

pontusmelke replied:
```
 /**
     * Flag to automatically compute the stack map frames of methods from
     * scratch. If this flag is set, then the calls to the
     * {@link MethodVisitor#visitFrame} method are ignored, and the stack map
     * frames are recomputed from the methods bytecode. The arguments of the
     * {@link MethodVisitor#visitMaxs visitMaxs} method are also ignored and
     * recomputed from the bytecode. In other words, computeFrames implies
     * computeMaxs.
     * 
     * @see #ClassWriter(int)
     */
    public static final int COMPUTE_FRAMES = 2;
```

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/TryBlock.java:
```diff
@@ -19,20 +19,165 @@
+            return parent;
+        }
+    }
+    private State newState( State parent )
```
This looks like it could be static. If this method is static the anonymous inner class will not keep a reference to instantiating instance of the surrounding class.

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/CodeBlock.java:
```diff
@@ -73,26 +87,31 @@ private void endBlock()
+    {
+        emitFunction.accept( emitter );
+    }
+
```
This indirection looks like it was added in order to defer the actions. It was used in `TryBlock`. It seems like then the overriding of this method has been removed, rendering this indirection unnecessary and costly (extra object creation, extra indirection).

pontusmelke replied:
yes, nice catch!

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/bytecode/ByteCodeExpressionVisitor.java:
```diff
@@ -0,0 +1,600 @@
+    public void equal( Expression lhs, Expression rhs, TypeReference type )
+    {
+        switch ( type.simpleName() )
+        {
```
Have we completely removed the notion of `Expression` carrying a type? It seems to me that the type parameter is a bit superfluous.

pontusmelke replied:
I don't think it was ever there, but it would be useful I agree.

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/ExpressionVisitor.java:
```diff
@@ -41,15 +41,37 @@
+
+    void longToDouble( Expression expression );
+
+    void pop( Expression expression );
```
is `pop(...)` really used? Looks like it was at some point but then the use was removed.

thobe replied:
The reason I went looking was because I didn't like the name...

The purpose of this method would be to evaluate the expression for it's side effects only, and we already have methods for that, on the _Statement_ level.

pontusmelke replied:
pop is used if we do stuff like

```
map.put("foo", "bar);
```

(where put returns a value we don't care about)

the generated byte code should be

```
LDC "foo"
LDC "bar"
INVOKEVIRTUAL java/util/HashMap.put (Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
POP
```

thobe replied:
Yes, exactly. I don't think `pop` is a great name for that. I also think that the `.expression( Expression )` should just do the popping, i.e. no need to have an expression that pops.

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/CodeBlock.java:
```diff
@@ -117,74 +137,82 @@ public Expression self()
     }
 
-    CodeBlock emitCatch( Parameter exception )
+    public CodeBlock ifNotStatement( Expression test )
```
Could we not have the `NotExpression` perform this optimization instead? So that the surface API still remains a single method for `if`.

pontusmelke replied:
Maybe somewhat of a premature optimization but I want to generate the same code as javac. So 

```
if (!flag) 
{
    System.out.println("hello");
}
```

will compile to 

```
  ILOAD 1
  IFNE L0
  GETSTATIC java/lang/System.out : Ljava/io/PrintStream;
  LDC "hello"
  INVOKEVIRTUAL java/io/PrintStream.println (Ljava/lang/String;)V
L0
 ...
```

whereas if we use a `NotExpression` we would have something like

```
  ILOAD 1
  ...invert boolean
  IFEQ L0
  GETSTATIC java/lang/System.out : Ljava/io/PrintStream;
  LDC "hello"
  INVOKEVIRTUAL java/io/PrintStream.println (Ljava/lang/String;)V
L0
    ...
```

The reason for the other specialized `ifX` are the same I want to use the opcodes `IFNULL`, `IFNONNULL` and not have to resort to compare against null.

thobe replied:
I agree we want to use those opcodes. I disagree with this being something we should surface at this level of the API. Instead I think we should specialise the _compare to null_ and _not_ `Expression` subtypes.

The `not` subtype should also specialise so that `not(not(expr))` evaluates to `expr`.

pontusmelke replied:
Currently `beginIf(expr)` generates

```
"LOAD expr"
IFEQ L0
```

If I understand you; `beginIf` should not generate the `IFEQ` but that should be handled by the Expression? Otherwise I don't see how we can generate the appropriate opcodes like `IFNE`, `IFNULL`, and `IFNONNULL`?

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/CodeBlock.java:
```diff
@@ -117,74 +137,82 @@ public Expression self()
     }
 
-    CodeBlock emitFinally()
+    public CodeBlock ifNullStatement( Expression test )
```
Same thing here as for `ifNotStatement`, could we not have the fact that the `Expression` is a null check fold to such an efficient comparison?

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/CodeBlock.java:
```diff
@@ -117,74 +137,82 @@ public Expression self()
     }
 
-    public CodeBlock tryBlock( Class<?> resourceType, String resourceName, Expression resource )
+    public CodeBlock ifNonNullStatement( Expression test )
```
Same thing here as for `ifNotStatement`, could we not have the fact that the `Expression` is a null check fold to such an efficient comparison?

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/Expression.java:
```diff
@@ -153,6 +237,30 @@ public void accept( ExpressionVisitor visitor )
         };
     }
 
+    public static Expression ternaryOnNull( final Expression test, final Expression onTrue, final Expression onFalse )
```
Same thing here as for `ifNotStatement`. Instead of having a special method for this ternary, could we not provide more insight into the `test` `Expression` in order to perform this optimisation in the implementation rather than forcing the API user to do it?

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/Expression.java:
```diff
@@ -153,6 +237,30 @@ public void accept( ExpressionVisitor visitor )
+        };
+    }
+
+    public static Expression ternaryOnNonNull( final Expression test, final Expression onTrue, final Expression onFalse )
```
Same thing here as for `ifNotStatement`. Instead of having a special method for this ternary, could we not provide more insight into the `test` `Expression` in order to perform this optimisation in the implementation rather than forcing the API user to do it?

thobe commented:
Overall very good changes. Sure there are a few things that could be better, but we could make those improvements after merging this, since overall these changes are a vast improvement over what we had before.

There is also a lot of cleanup and documentation I would like to see done. Quite a few methods should have short description about what their intention is. I also would like us to investigate if we can make the APIs nicer to work with now that Java 8 is the target.

pontusmelke commented:
@thobe I think we can remove a lot of stuff now that we are targeting java 8 (TypeReference, Parameters,... all have java 8 equivalents I think). My goal was to change as little as possible on the API level in order to keep the PR at least somewhat manageable.

thobe commented:
@pontusmelke I agree. Such changes are better done later.

thobe commented:
@pontusmelke let's leave `ifNot`, `ifNull`, et.c. for a later change and not hold up this PR because of that. Anything else we wish to change before merging? Because it looks good enough to me now.
